### DISCUSSION TITLE
Issue with Tableau operator to refresh workbook.

### DISCUSSION THREAD

**Discussion by hemantaggarwal** [apache/airflow#27766] [Q&A]
I am using AWS MWAA ver 2.2.2 and trying to use the Tableau operator to refresh a workbook. I created a Tableau connection in Airflow UI and provided the username and password there. In the Tableau operator, I passed the following parameters

`TableauOperator(
    task_id='refresh_tableau_workbook',
    resource='workbooks',
    method='refresh',
    find=workbook_name,
    match_with='name',
    site_id=site_id,
    tableau_conn_id=tableau_conn_id,
    dag=dag
)`

But when I run it I get the following error: **tableauserverclient.server.exceptions.NotSignedInError: Missing site ID. You must sign in first.**

If tableau online has "/#/site/test/projects/4" in the URL then I am passing the "test" to the site_id variable. I am not sure where else I am missing site_id.

**Comment by Taragolis** [comment]
Hey @hemantaggarwal the problem might be with `tableauserverclient` and integration Airflow Provider

This might be the same situation that happen with [this issue](https://github.com/apache/airflow/issues/24507#issuecomment-1158261674) and might refers to situation that actually Tableau  provider might not work correctly with new version of `tableauserverclient`.

I would recommend install `tableauserverclient==0.18.0` and try to check if error still persist.

**Reply by hemantaggarwal** [reply]
@Taragolis: I am using tableauserverclient version 0.17.0 as the constraint [file](https://raw.githubusercontent.com/apache/airflow/constraints-2.2.2/constraints-3.7.txt) contains this version only.

**Reply by Taragolis** [reply]
That mean my initial assumption is wrong 😞  and this is not your case.

You told that you use providers listed in constant files than I assume that you use `2.1.2` version of tableau provider.

I have a quick look on `airflow.providers.tableau.operators.tableau.TableauOperator` for selected version of provider and found that refresh operator executed out of the context manager

https://github.com/apache/airflow/blob/d9567eb106929b21329c01171fd398fbef2dc6c6/airflow/providers/tableau/operators/tableau.py#L110-L121

It is looks like `refresh` method called after connection closed. It fixed by PR https://github.com/apache/airflow/pull/20240, actually it fix static checks but also seems like it fix problem with connection. This changes included into `2.1.3` version of provider

https://github.com/apache/airflow/blob/f77417eb0d3f12e4849d80645325c02a48829278/airflow/providers/tableau/operators/tableau.py#L114-L132

So try to install `apache-airflow-providers-tableau==2.1.3` or even latest version of provider.

In case if error would persist could you also add exception traceback which might help to someone more familiar with this provider to help with troubleshooting.

**Reply by hemantaggarwal** [reply]
@Taragolis: How can I bypass the constraint for the newer version of apache-airflow-providers-tableau? Because in the constraint file the provider version is 2.1.2, due to which I cannot install ver 2.1.3.

**Reply by Taragolis** [reply]
I do not know how to do it within MWAA because you could provide only single requirements file in AWS MWAA.
So if you describe constraints in requirements file there is no way to disable it.

But this is info how to do it in general way
https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#installing-upgrading-downgrading-providers-separately-from-airflow-core

Also you might try to use `https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-no-providers-${PYTHON_VERSION}.txt`

**Reply by hemantaggarwal** [reply]
@Taragolis: Thanks a lot. Changing the constraint file to no-providers worked for me and I was able to install tableau provider version 2.1.3 and now the issue is also resolved. 
I will check with AWS if there is a way to use multiple requirements files if we use normal constraint files but for now, at least we have a working solution. 
I think as this conversation is in the comments, I can't select your last comment as an answer. Shall I select your first comment as the answer?

**Reply by potiuk** [reply]
@Taragolis - this is actually an interesting idea to use the "no-providers" one for such case. 

I have not thought about this use case but in fact this is something that might help to make the "core dependencies" intact for the version installed. While it might not help in some cases when there is actual conflict, it might prevent the users from accidentally upgrading Airlfow for example. I will add some docs for that before 2.1.5. I might even slightly improve the "no-providers" constraints to actually contain airflow in-there to be absolutely sure the accidental upgrades won't happen.

**Reply by Taragolis** [reply]
I had some internal investigation and found that this approach would not work in all cases.

I did not have an issue with Airflow 2.2 but had with Airflow 2.3.x + snowflake provider as I remember in `no-providers` cryptography constraint 37.x but snowflake at that time `snowflake-connector-python` required required <37.

So as solution we started generate own constraint every time we change some dependency in the project.

**Reply by hemantaggarwal** [reply]
So I checked with AWS, and they had a way to use a custom constraint file in MWAA (https://docs.aws.amazon.com/mwaa/latest/userguide/connections-packages.html#connections-packages-newer-packages). 
I tried this by changing the tableau provider version in the normal constraint file and this method also works fine. 
So I think, we will go with the custom/modified constraint file method.
